### PR TITLE
Add tests for stranded successors in DAGs (resolves #1122, resolves #1289, resolves #1294)

### LINKED ISSUE #1: Can Jobtree be extended to the set of requirements for this

### LINKED ISSUE #1122: cwltoil non-clean job failures
While debugging #1107 and responding to comments from @hannes-ucsc I realized the underlying issues with CWL restarts might be that cwltoil does not exit cleanly. Here is a small example with the test directory, forcing revtool to fail by messing up the `rev` command:

```
cd src/toil/test/cwl
sed -i 's/baseCommand: rev/baseCommand: rev-fail/' revtool.cwl
cwltoil --no-container revsort.cwl revsort-job.json
```

This fails with an assertion error during consistency checks at the end of `innerLoop` (https://github.com/BD2KGenomics/toil/blob/92f64a5d94bd16514979bc5c784f72863f2ddf5d/src/toil/leader.py#L728) which I suspect is not the right outcome. That function is pretty complex and deeply nested so it's not clear to me how to start debugging this. I hope that there's something we can include from the CWL side to create cleaner job exits here.

Can anyone offer any advice/directions?

Here is the full traceback from the run:

```
kunkel 2016-08-26 11:18:31,313 MainThread WARNING toil.leader: The jobWrapper seems to have left a log file, indicating failure: B/U/jobWBMw6I
kunkel 2016-08-26 11:18:31,313 MainThread WARNING toil.leader: Reporting file: B/U/jobWBMw6I
kunkel 2016-08-26 11:18:31,313 MainThread WARNING toil.leader: B/U/jobWBMw6I:   ---TOIL WORKER OUTPUT LOG---
kunkel 2016-08-26 11:18:31,313 MainThread WARNING toil.leader: B/U/jobWBMw6I:   INFO:rdflib:RDFLib Version: 4.2.1
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:   INFO:toil.fileStore:CACHE: Read file with ID 'a/2/tmp2RYy7z.tmp' from the cache.
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:   [job revtool.cwl] /tmp/user/1001/tmpZ7hLZ6$ rev-fail \
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:       /tmp/user/1001/tmppbMf7F/stg5c0b758e-c348-411b-b260-877b9491c6da/whale.txt > /tmp/user/1001/tmpZ7hLZ6/output.txt
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:   INFO:cwltool:[job revtool.cwl] /tmp/user/1001/tmpZ7hLZ6$ rev-fail \
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:       /tmp/user/1001/tmppbMf7F/stg5c0b758e-c348-411b-b260-877b9491c6da/whale.txt > /tmp/user/1001/tmpZ7hLZ6/output.txt
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:   'rev-fail' not found
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:   ERROR:cwltool:'rev-fail' not found
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:   [job revtool.cwl] completed permanentFail
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:   WARNING:cwltool:[job revtool.cwl] completed permanentFail
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:   Final process status is permanentFail
kunkel 2016-08-26 11:18:31,314 MainThread WARNING toil.leader: B/U/jobWBMw6I:   WARNING:cwltool:Final process status is permanentFail
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:   INFO:toil.fileStore:LOG-TO-MASTER: Successfully deleted local copies of file with ID 'a/2/tmp2RYy7z.tmp'.
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:   Traceback (most recent call last):
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:     File "/home/chapmanb/install/bcbio-vm/anaconda/lib/python2.7/site-packages/toil/worker.py", line 331, in main
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:       job._runner(jobWrapper=jobWrapper, jobStore=jobStore, fileStore=fileStore)
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:     File "/home/chapmanb/install/bcbio-vm/anaconda/lib/python2.7/site-packages/toil/job.py", line 1085, in _runner
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:       returnValues = self._run(jobWrapper, fileStore)
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:     File "/home/chapmanb/install/bcbio-vm/anaconda/lib/python2.7/site-packages/toil/job.py", line 1033, in _run
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:       return self.run(fileStore)
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:     File "/home/chapmanb/install/bcbio-vm/anaconda/lib/python2.7/site-packages/toil/cwl/cwltoil.py", line 256, in run
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:       **opts)
kunkel 2016-08-26 11:18:31,315 MainThread WARNING toil.leader: B/U/jobWBMw6I:     File "/home/chapmanb/install/bcbio-vm/anaconda/lib/python2.7/site-packages/cwltool/main.py", line 218, in single_job_executor
kunkel 2016-08-26 11:18:31,316 MainThread WARNING toil.leader: B/U/jobWBMw6I:       raise WorkflowException(u"Process status is %s" % (final_status))
kunkel 2016-08-26 11:18:31,316 MainThread WARNING toil.leader: B/U/jobWBMw6I:   WorkflowException: Process status is ['permanentFail']
kunkel 2016-08-26 11:18:31,316 MainThread WARNING toil.leader: B/U/jobWBMw6I:   Exiting the worker because of a failed jobWrapper on host kunkel
kunkel 2016-08-26 11:18:31,316 MainThread WARNING toil.leader: B/U/jobWBMw6I:   ERROR:toil.worker:Exiting the worker because of a failed jobWrapper on host kunkel
kunkel 2016-08-26 11:18:31,316 MainThread WARNING toil.leader: B/U/jobWBMw6I:   WARNING:toil.jobWrapper:Due to failure we are reducing the remaining retry count of job B/U/jobWBMw6I to 0
kunkel 2016-08-26 11:18:31,316 MainThread WARNING toil.leader: B/U/jobWBMw6I:   WARNING:toil.jobWrapper:We have increased the default memory of the failed job to 2147483648 bytes
kunkel 2016-08-26 11:18:31,316 MainThread WARNING toil.leader: Job: B/U/jobWBMw6I is completely failed
kunkel 2016-08-26 11:18:31,316 MainThread INFO toil.leader: No jobs left to run so exiting.
kunkel 2016-08-26 11:18:31,316 MainThread INFO toil.leader: Finished the main loop
kunkel 2016-08-26 11:18:31,316 MainThread INFO toil.leader: Waiting for stats and logging collator thread to finish ...
kunkel 2016-08-26 11:18:31,638 MainThread INFO toil.leader: ... finished collating stats and logs. Took 0.321724891663 seconds
kunkel 2016-08-26 11:18:31,639 MainThread INFO toil.leader: Waiting for service manager thread to finish ...
kunkel 2016-08-26 11:18:32,129 MainThread INFO toil.leader: ... finished shutting down the service manager. Took 0.490026950836 seconds
Traceback (most recent call last):
  File "/usr/local/bin/cwltoil", line 9, in <module>
    load_entry_point('toil==3.5.0a1', 'console_scripts', 'cwltoil')()
  File "/home/chapmanb/install/bcbio-vm/anaconda/lib/python2.7/site-packages/toil/cwl/cwltoil.py", line 665, in main
    outobj = toil.start(wf1)
  File "/home/chapmanb/install/bcbio-vm/anaconda/lib/python2.7/site-packages/toil/common.py", line 580, in start
    return self._runMainLoop(job)
  File "/home/chapmanb/install/bcbio-vm/anaconda/lib/python2.7/site-packages/toil/common.py", line 794, in _runMainLoop
    jobCache=self._jobCache)
  File "/home/chapmanb/install/bcbio-vm/anaconda/lib/python2.7/site-packages/toil/leader.py", line 692, in mainLoop
    innerLoop(jobStore, config, batchSystem, toilState, jobBatcher, serviceManager, statsAndLogging)
  File "/home/chapmanb/install/bcbio-vm/anaconda/lib/python2.7/site-packages/toil/leader.py", line 970, in innerLoop
    assert toilState.successorCounts == {}
AssertionError
```

### LINKED ISSUE #1289: Fix job scheduling on restart after partially failed DAGs
If a job in a DAG fails and there is another path to its child, the child is run in the restart irrespective of the status of the previously failed parent in the new run.

@arkal will write a test to prove this behaviour and wither @cket or @benedictpaten will probably write the fix.

### PR DESCRIPTION
**Pull request by cket**
@benedictpaten this adds a couple simple test cases for #1122

### CODE REVIEW

**Comment by hannes-ucsc** [comment]
A note on process. The commit that adds the tests should be "Add tests for stranded successors (connected to #1122)" as it doesn't resolve the issue, just exposes it. The commit that resolves the issue should be "… (resolves #1122)".

**Comment by cket** [comment]
@benedictpaten I just added another test case I thought I had already committed. I also disabled jenkins from running on this PR until we have something we want to test. To re-enable it you can just edit the PR's description to remove the 'jenkins' directive.

**Comment by benedictpaten** [comment]
Okay, I think I have a solution for this. The switching off asserts is not a general solution for, as it would allow deadlocks to occur.

**Comment by benedictpaten** [comment]
jenkins, test this please

**Comment by benedictpaten** [comment]
@hannes-ucsc Is something up with Jenkins?

**Comment by cket** [comment]
@benedictpaten no, I had temporarily disabled jenkins for the PR https://github.com/BD2KGenomics/toil/pull/1264#issuecomment-258552571

**Comment by cket** [comment]
jenkins, test this please

**Comment by benedictpaten** [comment]
Not sure what's going on with these test failures, @cket ?

**Comment by cket** [comment]
The hot deployment failures seem unrelated and maybe sporadic. The sort test and service test failures are probably related. I can take a look at debugging them if you want @benedictpaten

**Comment by benedictpaten** [comment]
I'll take a look, but you can too

Benedict
(from phone, forgive extra typos)

File: Makefile [hunk 1]
```diff
@@ -92,7 +92,7 @@ clean_sdist:
 	- rm -rf dist
 
 
-test: check_venv check_build_reqs docker
+test: check_venv check_build_reqs
```
Comment (hannes-ucsc) [inline Makefile]: This is the reason for the failed hot-deployment tests.

Comment (benedictpaten) [inline Makefile]: Ahh, but if I enable that it refuses to work on my laptop?

Comment (hannes-ucsc) [inline Makefile]: You can have that change locally, you just shouldn't have committed it.

**Comment by cket** [comment]
I added Arjun's test, it exposes 2 potential problems. The number of failed jobs is different than expected (but if I understand this PR that change is intentional), and the testRestartedWorkflowSchedulesCorrectJobsOnKilledParent still throws the assertion error while the testRestartedWorkflowSchedulesCorrectJobsOnFailedParent seems OK. Unfortunately the assertion's traceback is obscured by the test right now, I'll fix that right after lunch. 

Both tests indicate that the restart problem #1289 is fixed on this branch.

**Comment by benedictpaten** [comment]
Oops...

**Comment by benedictpaten** [comment]
@arkal 's test is still failing, not sure why but will fix. The rest is working on my end.

**Comment by benedictpaten** [comment]
I expect these tests to pass, once they do I'd propose merging this PR and closing #1289 #1294 #1292. The CWL restart issue PR #1107 can then be addressed.

**Comment by benedictpaten** [comment]
@hannes-ucsc @cket please review, would like to get this in asap so I can build on it.

**Comment by hannes-ucsc** [comment]
The commit message is wrong. It should have been

Resolves #1122, resolves #1289, incorporates #1292

**Pull request merged**